### ISSUE TITLE
^ symbol issues

### ISSUE THREAD

**Issue by ETLegacyTeam** [etlegacy/etlegacy#412]
Author: @saukko28 (Saukko)
Date: 2013-12-08
Redmine Issue: 412, https://dev.etlegacy.com/issues/412

---

Ok there are several issues that "&#94;" symbol creates:

- When you type &#94; once, it won't show anything in console. (see the 1st line on the picture)
- When you type &#94; once gain, it won't show anything in console and will put the text on the same line instead of a different line. (1st line)
- When you type &#94;&#94;, it will show only one &#94; in console. (2nd line)
- When you type &#94;&#94; another time it will only show it like this "&#94;]&#94;" (2nd line)
- When you type for example "&#94;sau" in console, the text will turn yellow and will in the end show au in yellow colour (Works as intended?)
- When you type for example "&#94;&#94;sau" in console, it will be the same as the one example above but puts green &#94; symbol infront of the text.

When you type in chat (T), then it will work way differently.

-*S

---

Related issue(s): #502

---

Related commit(s):
b550a1febe91e693d04323cc540f1b223d3137a2 - general: fixed inaccurate Q_IsColorString(), fixes..
 8be99042e965642603ec03a657543bb0aaf31c19 - client: fixed console color codes input, refs 412..
 0e4504a80289a26fbf1dd2e6c9eb188ea7bccd48 - client: fixes color code issues introduced recentl..

**Comment by ETLegacyTeam** [comment]
From: @saukko28 (Saukko)
Date: 2013-12-26T14:54:03Z

---

^ symbol works as intended in the intermission chat, you get one ^ and press "S" for example to get yellow colour. The ^ symbol fades away and the you can see is empty chat line where you can start typing with yellow colour.

In chat box (default "T") ^ symbol doesn't work as intended. Instead of ^ just fading away when you have pressed one, it stays. And when you start typing after the ^ symbol it would look like this for example: ^objective but in the end when you press Enter to submit the text, it will end up as "bjective" with the colour that "O" button gives.

So to make things work like supposed to, the chat box should be fixed to work like in the intermission chat when it comes to colours.

With Finnish keyboard style the ^ symbol is inconsistent and it has been confirmed. In vanilla ET it works differently since you have to press Shift+¨ twice to get two ^ symbols because once is not enough. So in other words it's a so called "dead key". But still it works correctly in the vanilla ET without any kind of malfunctions. ET:L in the other hand makes it work differently. When you press Shift+¨ once it gives one ^. Which is how it supposed to work. But it is inconsistent as mentioned.

-*S

**Comment by ETLegacyTeam** [comment]
From: @JanSimek (Radegast)
Date: 2014-01-13T02:07:17Z

---

Fixed by user:swillits in "this pull request":https://github.com/etlegacy/etlegacy/pull/75. Pending review by Ensiform.

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2014-01-13T10:02:28Z

---

Sorry, I commited the change before I saw that Ensi should review it first. Changing to "feedback".

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2014-01-13T10:15:49Z

---

Applied in changeset commit:b550a1febe91e693d04323cc540f1b223d3137a2.

**Comment by ETLegacyTeam** [comment]
From: @JanSimek (Radegast)
Date: 2014-01-13T16:29:58Z

---

No problem. He already did.

**Comment by ETLegacyTeam** [comment]
From: @saukko28 (Saukko)
Date: 2014-03-14T22:17:03Z

---

This thing isn't over yet. There are two things that are faulty now:

1. First are the two pictures showing how ^ creates a bug. The first picture shows the faulty text in the console and the second picture shows how the name should really be.

2. When you do ^^ in the console, it shows it as darkblue, and that is the reason why that makes it buggy.

Fixed to New?

-*S

**Comment by ETLegacyTeam** [comment]
From: @jackeri (Jacker)
Date: 2014-03-14T22:20:08Z

---

Lets check this out (again)

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2014-03-14T22:25:57Z

---

Dark blue is the correct behavior, since this is the intended color code of '^' (only enabled after a first '^' obviously).

**Comment by ETLegacyTeam** [comment]
From: @swillits (swillits)
Date: 2014-03-15T16:49:15Z

---

In a sane world, "^x" (where x is not "^") would be a color code, and "^^" would simply escape to a single literal "^" without any color change. This would be consistent with nearly every formatter and escaper syntax on the planet, but for reasons beyond me, it simply doesn't work that way. The code is trivially simpler by working the way it does, but the behavior is wacky. (And for backwards-compatability reasons, it can't change.)

**Comment by ETLegacyTeam** [comment]
From: @jackeri (Jacker)
Date: 2014-03-15T17:56:23Z

---

swillits wrote:
> In a sane world, "^x" (where x is not "^") would be a color code, and "^^" would simply escape to a single literal "^" without any color change. This would be consistent with nearly every formatter and escaper syntax on the planet, but for reasons beyond me, it simply doesn't work that way. The code is trivially simpler by working the way it does, but the behavior is wacky. (And for backwards-compatability reasons, it can't change.)

But isn't this exactly the way it works in the vanilla engine, and this is the way it should work atleast in my eyes. I need to fire up vanilla ET on my windows machine to test this but i am pretty sure that this is the way..

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2014-03-16T15:22:33Z

---

Jacker wrote:
> But isn't this exactly the way it works in the vanilla engine, and this is the way it should work atleast in my eyes. I need to fire up vanilla ET on my windows machine to test this but i am pretty sure that this is the way..

Just checked it, and it seems to work that way in vanilla ET. I thought my last change gave a similar behavior, but we're still far from it. The worst thing is that I don't know why the original behavior was changed in the first place (or was it to fix the newline character that was eaten?).

Subsequently, we added code to solve new bugs that were introduced by the new changes (possibility to show color in the uncolored string, then looping in the uncolored name as a workaround, then we had different characters displayed in the color and uncolored string, then we applied a ^^ that disable itself to fix this new issue...).

In a nutshell, all the fixes applied so far still result in an messy situation. Here's my suggestion: let's revert back everything to the original code (including the buggy new line), and try to fix this bug for real in 2.72.

**Comment by ETLegacyTeam** [comment]
From: @swillits (swillits)
Date: 2014-03-16T15:59:38Z

---

There's nothing to fix "for real." There were obvious bugs which caused multiple lines to render on one, and I fixed that. 

The color codes now work exactly as intended, without any bugs like "^\n" causing crazy issues.

The intended behavior is still weird, but that's just the way it is, and it can't be changed because it'd break many nicknames relying on the behavior that already exists. There's no point in changing that.


The only change list above other than the one I made, is commit:8be99042; I haven't studied it, but it doesn't look 'messy' to me. So I'm not sure what issue you're seeing with the current state?

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2014-03-16T17:39:49Z

---

Keep in mind the changes done in #502 too.

Also, vanilla ET doesn't work exactly as your "sane world" situation above, but allows to use '^' as a displayed character nonetheless (ie, @^1Spy^7^^7hawk@ would display @Spy^hawk@ with the 3 first letters red, then white letters). Unfortunately, the current code doesn't allow to use '^' as a displayed character (as reported by Saukko) and thus breaks nickname :(

It also seems that vanilla ET behavior has been broken way before swilits introduced the fix for the new line character bug. I'm actually unsure what is the change that broke it, and the reason of that change.

To summarize, the ideal color code behavior:
- should color strings exactly the same way as in vanilla ET,
- should display the same way in-game and in console,
- shouldn't break newline character in console,
- shouldn't allow in-game uncolored strings to display color (ie, @some^^78text@ should display @some^8text@ in white),
- should display colored and uncolored strings with the exact same characters (no difference, apart from the color),
- should display entered color codes in console (unlike vanilla ET) if possible.

Current code doesn't comply with first point (but comply with the others).
Last point is nice to have, but is the least important imho.

**Comment by ETLegacyTeam** [comment]
From: @swillits (swillits)
Date: 2014-03-16T17:58:17Z

---

> Unfortunately, the current code doesn't allow to use '^' as a displayed character (as reported by Saukko) and thus breaks nickname :(

It certainly does when I try it. Looking at his screenshots he never said what the original "input" text exactly was, so there's an assumption of what his original text was.

However to match the appearance of "therightone.jpg", the text should be "^^7V^0alk^7Y^0r^1!^0a" and if I plug that it into etlegacy right now, it works fine. I see it exactly as I expected, and exactly as it's shown in therightone.jpg. So, if there is a bug, I still don't know what it is. 

I just found this color code generator:
http://www.teammuppet.eu/home/index.php?/page/color-name-generator#%5E%5E7V%5E0alk%5E7Y%5E0r%5E1!%5E0a

It confirms the behavior I'm seeing.

**Comment by ETLegacyTeam** [comment]
From: @swillits (swillits)
Date: 2014-03-16T21:33:42Z

---

Ok, ignore my previous message. I had an old build without spyhawk's recent changes ref'd by 502.

I've reverted some of spyhawk's changes which caused the issues to pop back up, and more accurately fixed the display of names at the crosshairs, which was what he was after to begin with. 

commit:0e4504a80289a26fbf1dd2e6c9eb188ea7bccd48

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2014-03-18T18:28:25Z

---

Thx swillits. I haven't had time to test the changes yet, but looking closer at the code, I'm wondering now why we don't force color using CG_Text_Paint_Ext() a bit below instead of rewriting an entire new function Q_ColorizeString()?

Might require some adjustment, but it is probably the cleanest way to do it, isn't it?

Edit: After an IRC talk with swillits, it seems this isn't possible easily nor would this be desirable. The old drawing code (blurry fonts) had an extra parameter to do this, but we probably don't want to bring it back.
